### DISCUSSION TITLE
How to detect if a user has not created a password, or has only signed up through magic link

### DISCUSSION THREAD

**Discussion by briankelleher** [supabase/supabase#5741] [Questions]
Right now, it seems impossible to detect if a user has created a password on their account, or has only signed up through a "magic link".

If a user signs up via a magic link, there is a random encrypted password assigned to the user in the auth.users table.  Thus, it is impossible to detect if the user has set that password or not.

Additionally, you cannot confirm if a user has signed up via magic link or password if Email Confirmations are off - if they are off the confirmed_at field on auth.users is defaulted to the account creation date, which is misleading because the account has never been confirmed or used at the time of submission.

Is this on the developer to create a separate user logins table to determine these factors? Seems unnecessary, but maybe I'm missing something.

**Comment by liaujianjie** [comment]
Seems like the password is coming from this bit here where new signups via magic links are created with a temporary password:

https://github.com/supabase/gotrue/blob/bb7f120c39e650cd67e1a2fbf3565f33f9cafeac/api/magic_link.go#L47-L75

It also seems that the temporary password is not necessary for signing up via magic link. I was able to complete the magic link sign up flow by:
1. requesting for magic link
2. setting `encrypted_password` as an empty string for that user
3. clicking on magic link

**Reply by mb21** [reply]
@liaujianjie 

1. How do you request a magic link? with `supabase.auth.signInWithOtp()`? that unfortunately doesn't return an id... but yeah, we know the `email` that was used...
2. how do you set the `encrypted_password` to an empty string? just writing to the database directly or can this somehow be done with the gotrue js sdk?

**Comment by laygir** [comment]
Arrived here for similar question, I would like to invite users and remind them to set a password for their account after they followed the link on the email and landed on my page. Otherwise probably next time they want to login, they will scramble around to find the password they think they set but in actuality they never did. But can't find any piece of data point to get a hold off to rely when prompting for a new password.

**Comment by zubair1024** [comment]
Is there any concrete answer for this? Landed here as well. It should be a feature that should be there part of Supabase API honestly.

**Comment by TYRANTHOSAURUS** [comment]
Same issue, this is really needed to provide a smooth UX after a user has initially joined by magic link. The user is currently forced to continue using this method and request links instead of being capable of setting his password after using the original magic link.

**Comment by cipriancaba** [comment]
Hmm, really surprised to see that the default behaviour of supabase doesn't involve setting a password.  I want to force the users to set a password but as mentioned by OP it's not currently possible to detect this situation. This is very frustrating

Later edit: Just tested this and it seems that it actually works as expected. The `encrypted_password` field is actually empty after inviting users via magic link

**Reply by mb21** [reply]
really? how do you send the magic link? I just tested and using the js sdk's `supabase.auth.signInWithOtp()`, the `encrypted_password` field is not empty.

**Reply by cipriancaba** [reply]
Just did it from the supabase UI, sending the invite -> account gets created -> encrypted_password is empty

You could also set a value in the user_metadata which remembers if the user set the password, eg:

```
const resetPasswordRequest = await supabase.auth.updateUser({
      password: password,
      data: {
        hasPassword: true,
      },
    })
```

And then check in your code:
```
{!data.supaUser?.user_metadata.hasPassword ? (
        <RedirectToChangePassword />
      ) : ...
```

**Reply by RukshanJS** [reply]
I think this is a good approach.

**Reply by leomelzer** [reply]
Thanks @cipriancaba -- went with that idea, too.

**Comment by sasa-djuric** [comment]
I used the RPC function to detect whether the user has a password.

```sql
create function user_has_password(user_id uuid) returns boolean security definer as $$ declare retval bool;
begin
  select
    case 
      when encrypted_password is null or encrypted_password = ''
        then false 
      else true 
      end as has_password
  from
    auth.users into retval
  where
    id = user_id;
  return retval;
end;
$$ language plpgsql;
```

**Comment by mike-ayres** [comment]
The only time I see the `encrypted_password` column being empty is when oauth is used. Magic link registration has a value.

I went a similar route  [to this](https://github.com/orgs/supabase/discussions/5741#discussioncomment-7843834). When generating a magic link email, I add a bit of metadata to say magic link was used. We have our own user table with a trigger on `auth.users`; If the record is inserted, it looks at the magic link metadata I added and uses that to determine if a password has been set. But only on insert, so I don't have to clean up nodes on the raw metadata (at least not yet).

**Comment by MiroslavPetrik** [comment]
You can check for EmailProvider in the user's identities:

```ts
  const supabase = await createClient();

  const {
    data: { user },
  } = await supabase.auth.getUser();

  const hasPassword = user?.identities?.some((i) => i.provider === "email");
```

Edit: this does not work, because when I have a SSO user & create a password with the `.updateUser({password})` call, there won't be a new EmailIdentity created for the user...
